StackOverflowError when creating a policy with duty within its permissions

# Bug Report

## Describe the Bug
 In Policy creation functionality it fails to create a policy when within its odrl:permission a odrl:Permission contains a odrl:Duty.

### Expected Behavior
Create the odrl:Policy and stored in the database. 

### Observed Behavior
Throws Stack Overflow exception.

## Steps to Reproduce

1. Send POST to the endpoint "data/v2/policydefinitions"  with the following  odrl:Policy in the body:
```
{
  "@context": {
    "edc": "https://w3id.org/edc/v0.0.1/ns/",
    "cc": "http://creativecommons.org/ns#",
    "dalicc": "https://dalicc.net/ns#",
    "dalicclib": "https://dalicc.net/licenselibrary/",
    "dcat": "http://www.w3.org/ns/dcat#",
    "dct": "http://purl.org/dc/terms/",
    "foaf": "http://xmlns.com/foaf/0.1/",
    "odrl": "http://www.w3.org/ns/odrl/2/",
    "osl": "http://opensource.org/licenses/",
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
    "scho": "http://schema.org/",
    "xsd": "http://www.w3.org/2001/XMLSchema#"
  },
  "edc:policy": {  
    "@type": "odrl:Set",
    "odrl:permission": [
      {
        "@type": "odrl:Permission",
        "odrl:action": {
          "@id": "odrl:display"
        }
      }
    ],
    "odrl:prohibition": {
      "@type": "odrl:Prohibition",
      "odrl:action": {
        "@id": "dalicc:promote"
      }
    },
    "odrl:target": {
      "@type": "odrl:AssetCollection",
      "dct:type": [
        "dalicc:CreativeWork",
        "http://purl.org/dc/dcmitype/Dataset",
        "http://purl.org/dc/dcmitype/Software"
      ]
    }
}

```
throws `StackOverflowError`, 

```
SEVERE 2024-02-06T14:21:14.805513632 JerseyExtension: Unexpected exception caught
org.eclipse.edc.spi.persistence.EdcPersistenceException: com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion (StackOverflowError) (through reference chain: org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]-
[....]
>java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"])
	at org.eclipse.edc.connector.store.sql.policydefinition.store.SqlPolicyDefinitionStore.lambda$insert$5(SqlPolicyDefinitionStore.java:154)
	at org.eclipse.edc.transaction.local.LocalTransactionContext.lambda$execute$0(LocalTransactionContext.java:57)
	at org.eclipse.edc.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:74)
	at org.eclipse.edc.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:56)
	at org.eclipse.edc.connector.store.sql.policydefinition.store.SqlPolicyDefinitionStore.insert(SqlPolicyDefinitionStore.java:137)
	at org.eclipse.edc.connector.store.sql.policydefinition.store.SqlPolicyDefinitionStore.lambda$create$2(SqlPolicyDefinitionStore.java:101)
	at org.eclipse.edc.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:74)
	at org.eclipse.edc.connector.store.sql.policydefinition.store.SqlPolicyDefinitionStore.create(SqlPolicyDefinitionStore.java:97)
	at org.eclipse.edc.connector.service.policydefinition.PolicyDefinitionServiceImpl.lambda$create$7(PolicyDefinitionServiceImpl.java:107)
	at org.eclipse.edc.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:74)
	at org.eclipse.edc.connector.service.policydefinition.PolicyDefinitionServiceImpl.create(PolicyDefinitionServiceImpl.java:106)
	at org.eclipse.edc.connector.api.management.policy.PolicyDefinitionApiController.createPolicyDefinition(PolicyDefinitionApiController.java:109)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:146)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:189)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:93)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:261)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:240)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:697)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:357)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1381)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1303)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:192)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: org.eclipse.edc.spi.persistence.EdcPersistenceException: com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion (StackOverflowError) (through reference chain: org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]-
[...]
>java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"])
	at org.eclipse.edc.sql.store.AbstractSqlStore.toJson(AbstractSqlStore.java:84)
	at org.eclipse.edc.connector.store.sql.policydefinition.store.SqlPolicyDefinitionStore.lambda$insert$5(SqlPolicyDefinitionStore.java:143)
	... 60 more
Caused by: org.eclipse.edc.spi.persistence.EdcPersistenceException: com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion (StackOverflowError) (through reference chain: org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]-
[...]
>java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"]->java.util.ArrayList[0]->org.eclipse.edc.policy.model.Duty["parentPermission"]->org.eclipse.edc.policy.model.Permission["duties"])

```

## Context Information

- Used version EDC v0.2.1

## Possible Implementation

I believe the problem lies in the fact that both Permissions shows to Duties as well as Duty show back to parent Permission
https://github.com/eclipse-edc/Connector/blob/v0.2.1/spi/common/policy-model/src/main/java/org/eclipse/edc/policy/model/Duty.java#L47 

The problem probably remains in release 0.5.0, as the code for Duty class is the same:
https://github.com/eclipse-edc/Connector/blob/v0.5.0/spi/common/policy-model/src/main/java/org/eclipse/edc/policy/model/Duty.java#L47

--- a/core/common/lib/policy-engine-lib/src/main/java/org/eclipse/edc/policy/engine/ScopeFilter.java
+++ b/core/common/lib/policy-engine-lib/src/main/java/org/eclipse/edc/policy/engine/ScopeFilter.java
@@ -97,7 +97,6 @@ public class ScopeFilter {
         return Duty.Builder.newInstance()
                 .action(duty.getAction())
                 .constraints(filteredConstraints)
-                .parentPermission(duty.getParentPermission())
                 .consequences(filteredConsequences)
                 .build();
     }

--- a/spi/common/policy-model/src/main/java/org/eclipse/edc/policy/model/Duty.java
+++ b/spi/common/policy-model/src/main/java/org/eclipse/edc/policy/model/Duty.java
@@ -19,7 +19,6 @@ import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonTypeName;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
-import org.jetbrains.annotations.Nullable;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -33,23 +32,10 @@ import static java.util.stream.Collectors.joining;
 @JsonTypeName("dataspaceconnector:duty")
 public class Duty extends Rule {
 
-    private Permission parentPermission;
     private final List<Duty> consequences = new ArrayList<>();
-
-    /**
-     * If this duty is part of a permission, returns the parent permission; otherwise returns null.
-     */
-    @Nullable
-    public Permission getParentPermission() {
-        return parentPermission;
-    }
 
     public List<Duty> getConsequences() {
         return consequences;
-    }
-
-    void setParentPermission(Permission permission) {
-        parentPermission = permission;
     }
 
     @Override
@@ -72,11 +58,6 @@ public class Duty extends Rule {
         @JsonCreator
         public static Builder newInstance() {
             return new Builder();
-        }
-
-        public Builder parentPermission(Permission parentPermission) {
-            rule.parentPermission = parentPermission;
-            return this;
         }
 
         public Builder consequence(Duty consequence) {

--- a/spi/common/policy-model/src/main/java/org/eclipse/edc/policy/model/Permission.java
+++ b/spi/common/policy-model/src/main/java/org/eclipse/edc/policy/model/Permission.java
@@ -60,15 +60,11 @@ public class Permission extends Rule {
         }
 
         public Builder duty(Duty duty) {
-            duty.setParentPermission(rule);
             rule.duties.add(duty);
             return this;
         }
 
         public Builder duties(List<Duty> duties) {
-            for (var duty : duties) {
-                duty.setParentPermission(rule);
-            }
             rule.duties.addAll(duties);
             return this;
         }

